src/charm.py

```diff
@@ -53,6 +53,12 @@ def set_pod_spec(self, event):
         env = Environment(
             loader=FileSystemLoader("src/templates/"),
         )
+
+        if config["istio-gateway"] != "":
```

Comment:
By default, this is a non-empty [string](https://github.com/canonical/seldon-core-operator/blob/fix-istio-relation/config.yaml#L118), and I don't see where `config["istio-gateway"]` is emptied (unless I'm missing something). 

I would like to understand your decision of not using a relation here. I probably want to see a relation updating `ISTIO_ENABLED` rather than a config definition doing so. I see some scenarios where a relation could handle this better: istio could not even be deployed, `default-gateway` can point to something else, and there's nothing ensuring the gateway exists and is used by istio.

Reply:
Sorry for the slow response, I didn't notice this comment!

You're right, it is non-empty by default (but someone could do `juju config istio-gateway=""`) so I think it still needs handling.  In hindsight, I think I should also change the default value to `""` as I doubt many people by default will happen to have a gateway called `seldon-gateway` in the `istio-system` namespace.  Assuming we keep the design this way (instead of change it as discussed next), what do you think about switching default to `""`?  Also missing in the charm currently is any validation that this `Gateway` exists - that's probably something that should be added (eg: `lightkube.get()` for `Gateway` object of correct name, Block if missing).  Thoughts?

You're totally right about how istio might not be available, etc. - I don't like that either... At first I leaned toward a relation, but then went this way because I couldn't think of a relation that models this well.  Some ideas were:
* relation with istio-gateway: that isn't enough.  Istio-gateway deploys the workload that behaves as a gateway, but does not govern the `networking.istio.io/Gateway` object (that's created by istio-pilot).  
* relation with istio-pilot: that could be enough as it generates `Gateway` objects, but we'd need to add something to istio-pilot.  The `ingress` relation is not a good fit (it creates `VirtualServices`, which we explicitly do not want as Seldon will do that for itself, and does not return the `Gateway` name).  We could create a new `provided` relation in `istio-pilot` that provides the default-gateway name, and then relate `seldon:default-gateway->istio-pilot:default-gateway` as a replacement for the config.  That works so long as everything goes through the single default gateway, but we've talked about supporting multiple gateways.  If we supported multiple gateways, we could have a similar `gateways` relation that returns the list of available gateways, but then the seldon charm would still need some config to choose which gateway to actually use from that list (and we're back to the current implementation, except that we've got an additional relation).  

What do you think about this, did you have something specific in mind?  I'd love some help :)

Reply:
I get your point. Right now there is not a suitable interface that either of the istio charms provide, meaning we'll have to add it if we want to use a relation. I do not dislike the idea, we just have to be very careful where we add it and why. I didn't have something in mind when reviewing this PR, btw.

That said and based on the options you gave, I think a relation with `istio-pilot` could be our best option. As you said, `pilot` is the one actually handling `Gateway` and `VirtualService`, meaning it can create/destroy them as we relate or change the config, and is aware of them. We could do something like what you said:

> If we supported multiple gateways, we could have a similar gateways relation that returns the list of available gateways, but then the seldon charm would still need some config to choose which gateway to actually use from that list

What we get from that is that we are actually ensuring the gateway we need is there and is managed by the istio charm, in this case pilot, and not just assuming is there as we are doing right now.

Reply:
Would this relation actually ensure the gateway is there?  I don't see how it would.  If we implement a relation that lets the `Seldon` relate to `istio-pilot` to get the available `Gateway`s then the `Seldon` **charm** knows what is available, but the user still sets the desired gateway via `juju config seldon gateway-to-use=xyz` without that relation information.  With this new relation, the config is still a "try to use gateway `xyz`" (just like it is now).

Having the additional relation means that Seldon could raise an error saying "Configured gateway `xyz` not in gateways provided by istio-pilot relation".  But if that's what we're after, it might be easier to just get Seldon to look for the gateway we request directly (`lightkube.client.get(Gateway, name="xyz")`).